known_hosts: support --diff

mgedmin opened the pull request:
##### ISSUE TYPE
 - Feature Pull Request

##### COMPONENT NAME
known_hosts

##### ANSIBLE VERSION

ansible 2.3.0

##### SUMMARY

This makes `known_hosts` return a diff when invoked with `--check --diff`.

mgedmin committed:
known_hosts: support --diff

mgedmin commented:
I've tested this manually and it helped me figure out why ansible 2.3 reports `changed` for a couple of tasks in my playbook where ansible 2.2.1 didn't.

mgedmin committed:
known_hosts: support --diff also without --check

robinro commented on lib/ansible/modules/system/known_hosts.py:
```diff
@@ -119,12 +119,15 @@ def enforce_state(module, params):
 
     found,replace_or_add,found_line,key=search_for_host_key(module,host,key,hash_host,path,sshkeygen)
 
+    params['diff'] = compute_diff(path, found_line, replace_or_add, state, key)
```
Imo `params` is not the right place for this. Since you once use the content here, why not call it `diff`?

mgedmin replied:
This function returns `params`, which is then passed to `module.exit_json()`.  I need to set `params['diff']` to make the diff possible when `check_mode` is `False`.

bcoca replied:
i agree params is not the correct thing to use for return data .. but it would a separate fix for the module

robinro commented on lib/ansible/modules/system/known_hosts.py:
```diff
@@ -294,6 +297,30 @@ def normalize_known_hosts_key(key):
+        inf.close()
+    diff['after'] = ''.join([
+        line for line_number, line in enumerate(diff['before'].splitlines(1))
+        if not (found_line==(line_number + 1) and (replace_or_add or state=='absent'))
```
That logic is tough to understand. I'd rather make it more linear (code untested):
```
old_lines = diff['before'].splitlines()
if replace_or_add or state=='absent':
    del old_lines[found_line]
diff[after] = '\n'.join(old_lines)
```

mgedmin replied:
Note: found_line is 1-based, so it should be `del old_lines[found_line - 1]`.

I'm reluctant to do refactorings without unit tests.  I'll see if I can manage to write one.

mgedmin replied:
The tests discovered one bug (the diff would show the key being added even when no changes would be done), so yay!

robinro reviewed (changes requested):
Thanks @mgedmin for this PR. I'm looking forward to using this.

To ensure maintainability it's probably a good idea to cover this in the integration tests.
Add/remove a key in check_mode and ensure the diff looks as expected.

mgedmin committed:
Add unit tests and fix incorrect diff in one corner case

Tests are good!

mgedmin committed:
Refactor for readability

mgedmin committed:
Python 3 compat

mgedmin committed:
More Python 3 compat

mgedmin committed:
Add an integration test for known_hosts

mgedmin commented:
I've added an integration test (my first! not counting the small changes I've made to the authorized_diff integration test).

ready_for_review

mattclay commented on test/integration/targets/known_hosts/aliases:
```diff
@@ -0,0 +1,2 @@
+needs/root
```
Does this integration test really require root?

mgedmin replied:
I don't know!  I was copying what the authorized_key target did, without understanding what these things mean.

mattclay replied:
You can run the test locally as a non-root user to see if it passes (make sure you remove the `needs/root` alias first):

`test/runner/ansible-test integration known_hosts -v`

mgedmin replied:
I'm geting a weird failure if I do that:

```
TASK [known_hosts : add the same host in check mode] ***************************
task path: /home/mg/src/ansible/test/integration/targets/known_hosts/tasks/main.yml:63
fatal: [testhost]: FAILED! => {
    "changed": false, 
    "failed": true, 
    "invocation": {
        "module_args": {
            "hash_host": false, 
            "key": "example.org ssh-rsa AAAAB3NzaC1yc2EAAAADAQABAAAAglyZmHHWskQ9wkh8LYbIqzvg99/oloneH7BaZ02ripJUy/2Zynv4tgUfm9fdXvAb1XXCEuTRnts9FBer87+voU0FPRgx3CfY9Sgr0FspUjnm4lqs53FIab1psddAaS7/F7lrnjl6VqBtPwMRQZG7qlml5uogGJwYJHxX0PGtsdoTJsM=\n", 
            "name": "example.org", 
            "path": "/home/mg/ansible_testing/known_hosts", 
            "state": "present"
        }, 
        "module_name": "known_hosts"
    }, 
    "msg": "ssh-keygen failed to hash host (rc=1,stdout='|1|cM06NS72gSewxKexCR+837GhoYU=|ht5OEhWMLmnhMLN2QJvKkSrlvjE= ssh-rsa AAAAB3NzaC1yc2EAAAADAQABAAAAglyZmHHWskQ9wkh8LYbIqzvg99/oloneH7BaZ02ripJUy/2Zynv4tgUfm9fdXvAb1XXCEuTRnts9FBer87+voU0FPRgx3CfY9Sgr0FspUjnm4lqs53FIab1psddAaS7/F7lrnjl6VqBtPwMRQZG7qlml5uogGJwYJHxX0PGtsdoTJsM=\n',stderr='')"
}
```

Which is, incidentally, the same failure I see on Shippable.

Hmm.

mgedmin committed:
Handle ssh-keygen -HF returning non-zero exit code

AFAICT this is a bug in ssh-keygen in some newer OpenSSH versions
(>= 6.4 probably; see commit dd9d5cc670eccf3c92c8bf974dd294787fe94169):
when you invoke ssh-keygen with -H and -F <host> options, it always
returns exit code 1.  This is because in ssh-keygen.c there's a function
do_known_hosts() which calls

  exit (find_host && !ctx.found_key);

at the end, and find_host is 1 (because we passed -F on the command line),
but ctx.found_key is always 0.  Why is found_key always 0?  Because the
callback passed to hostkeys_foreach(), which is known_hosts_hash(),
never bothers to set found_key to 1.

mgedmin committed:
This test does not need root

mgedmin commented:
Fixed some Shippable failures, still have others.  E.g. `ssh-keygen` on CentOS 6 produces

```
line 5 invalid key: example.net...
/root/ansible_testing/known_hosts is not a valid known_hosts file.
```

because the old OpenSSH doesn't like `ssh-ed25519` keys, I think.

mgedmin committed:
Avoid ssh-ed25519 keys in sample known_hosts file

Older versions of OpenSSH do not like them and ssh-keygen -HF
aborts with an error when it sees such keys:

  line 5 invalid key: example.net...
  /root/ansible_testing/known_hosts is not a valid known_hosts file.

mgedmin committed:
Fix Python 3 errors

Specifically, the default mode of tempfile.NamedTemporaryFile is 'w+b',
which means Python 3 wants us to write bytes objects to it -- but the
keys we have are all unicode strings.

mgedmin commented:
Rebased on current devel (partially because I messed up one of the refactoring commits and left a huge four-line one-liner of dead code).

mgedmin commented:
The `dnf` failure on fedora25 is not my fault, honest!

ready_for_review

bcoca commented on lib/ansible/modules/system/known_hosts.py:
```diff
@@ -145,7 +148,7 @@ def enforce_state(module, params):
                                      (path,str(e)))
         try:
-            outf=tempfile.NamedTemporaryFile(dir=os.path.dirname(path))
+            outf = tempfile.NamedTemporaryFile(mode='w+', dir=os.path.dirname(path))
```
many changes that don't have to do with diff

mgedmin replied:
The reason those changes are there:
- I was asked to write an integration test
- the integration test uncovered some minor buglets in the code
- I had to fix those buglets to make the new integration test pass

Incidentally, the extra buglets were these two:
- the known_hosts module didn't work under Python 3 (not a blocker, since modules aren't expected to work under Python 3, but I didn't like the yellow blob in Shippable and the fix wasn't large)
- the known_hosts module failed with an error if you specified `hash_host=true` and had a relatively recent OpenSSH version (6.4 or newer) -- looks like this was already filed as bug #20500.

Would you like me to extract the integration test and the bugfixes into a separate PR?  I would prefer not to (it's quite a bit of extra work for me), but I can be easily convinced with a little bit of blackmail (e.g. "we won't merge this unless you do that").

bcoca replied:
no need for separate, just needed that explanation

mgedmin replied:
I keep writing extensive commit messages and then forgetting reviewers have to take extra steps to see them in GitHub's review UI.

(And when I'm reviewing PRs for projects I maintain, I tend to ignore commit messages and just look at the final diff.)

bcoca replied:
I normally rebase and put all messages in main commit (which shows in summary by default), but when you are doing several iterations its OK as is. When merging, we normally squash them if you hadn't already.

bcoca commented on lib/ansible/modules/system/known_hosts.py:
```diff
@@ -240,7 +243,7 @@ def search_for_host_key(module,host,key,hash_host,path,sshkeygen):
     sshkeygen_command.insert(1,'-H')
     rc,stdout,stderr=module.run_command(sshkeygen_command,check_rc=False)
-    if rc!=0: #something went wrong
+    if rc not in (0, 1) or stderr != '': #something went wrong
```
why is 1 allowed? afaik this indicates an error in ssh-keygen

mgedmin replied:
Yes, this is a bug in ssh-keygen.  The commit message of 5eef669313594c809dfe66f27786618409ed91a0 has all the gory details.

mgedmin replied:
BTW https://github.com/ansible/ansible/commit/dd9d5cc670eccf3c92c8bf974dd294787fe94169 was a workaround for a very similar issue.  And then https://github.com/ansible/ansible/commit/405c6820b0afd74fd335d2d302028e7e816f5bb9 added a 2nd invocation of ssh-keygen, but was apparently never tested with newer OpenSSH versions.

bcoca replied:
**sigh** i really hate compensating for these types of bugs .... but I don't see we have much of a choice. tempted to add a warning "your ssh is broken".

mgedmin replied:
I feed kind of bad that I didn't report the bug upstream, but "OpenSSH" makes me think "maintained by OpenBSD guys", which leads to "probably in some CVS repository from 1998 with a mailing list instead of a bug tracker", and I don't have the energy.

(Also I'm probably wrong and they use Git, or at least Subversion now.  Still.)

bcoca replied:
FYI, https://www.openssh.com/report.html, they use bugzilla.

mgedmin commented:
What's the status of this?  @bcoca's review indicates "requested changes", but it's not clear to me what changes are being requested.

mgedmin commented:
:sparkles: Thank you!  :sparkles:
